### PR TITLE
[Feature] Migrate KDA Triton Op to Ascend NPU

### PR DESCRIPTION
**Pull request by yongfuFang**
### What this PR does / why we need it?

This PR migrates the KDA (Kimi Delta Attention) Triton operators from vLLM's upstream FLA (flash-linear-attention) GPU implementation to Ascend NPU. Previously, KDA had no NPU-compatible Triton implementation.

### Does this PR introduce _any_ user-facing change?

No new user-facing configuration or public API is introduced. This PR only adds the underlying Ascend NPU Triton kernels for KDA, without changing existing behavior.

#### File responsibilities

| File | Responsibility | Main caller |
| --- | --- | --- |
| `kda.py` | Public operator entry points and orchestration for `chunk_kda`, `fused_recurrent_kda`, `fused_kda_gate`, and gated RMSNorm; also contains the chunk KKT, `W/U`, and output kernels. | Model integration / tests |
| `fused_recurrent_kda.py` | Single-kernel token-by-token KDA recurrence used by decode, including optional Q/K L2 normalization and paged in-place state updates. | `fused_recurrent_kda_fwd` in `kda.py` |
| `cumsum.py` | Computes the per-chunk cumulative vector gate `G`. | `chunk_kda_fwd` in `kda.py` |
| `solve_tril.py` | Computes the blockwise inverse `M = (I + L)^{-1}` for 16/32/64-token lower-triangular blocks. The current KDA path uses a chunk size of 64. | `chunk_kda_fwd` in `kda.py` |
| `chunk_delta_h.py` | Propagates the recurrent state across chunks and computes the corrected value block `V'`. | `chunk_kda_fwd` in `kda.py` |
| `l2norm.py` | Optional standalone Q/K L2 normalization for the chunk path. | `chunk_kda` in `kda.py` |
| `utils.py` | Cached variable-length metadata (`chunk_indices` and `chunk_offsets`) plus contiguous-input/device guards. | `kda.py`, `cumsum.py`, `solve_tril.py`, `chunk_delta_h.py` |

##### Prefill / chunked path
img

##### Decode / recurrent path
img_1

`fused_kda_gate` and `rms_norm_gated` are independent helper entry points in `kda.py`; they are not called from the two attention paths above. `fused_kda_gate` launches `kda_gate_fwd_kernel`, while `rms_norm_gated` selects one of the two `layer_norm_gated_fwd_kernel` variants according to the feature dimension.

#### Recurrent KDA definition

For readability, let the mathematical state be $S_t \in \mathbb{R}^{K \times V}$. The Triton kernels physically store its transpose, $S_t^\top \in \mathbb{R}^{V \times K}$, to make the value dimension contiguous. With $s = K^{-1/2}$ and optional

$$
q_t \leftarrow \frac{q_t}{\sqrt{\lVert q_t \rVert_2^2 + \epsilon}},
\qquad
k_t \leftarrow \frac{k_t}{\sqrt{\lVert k_t \rVert_2^2 + \epsilon}},
$$

the decode kernel implements the following recurrence in one launch:

$$
\widetilde{S}_t = \mathrm{Diag}(\exp(g_t)) S_{t-1},
$$

$$
\delta_t = \beta_t \odot \left(v_t - \widetilde{S}_t^\top k_t\right),
$$

$$
S_t = \widetilde{S}_t + k_t \delta_t^\top,
\qquad
o_t = s S_t^\top q_t.
$$

Here, $g_t \in \mathbb{R}^{K}$ is a key-dimension decay gate. The recurrent kernel accepts either a scalar $\beta_t$ or a value-dimension vector $\beta_t \in \mathbb{R}^{V}$. The `ssm_state_indices` path maps each sequence/token to its paged state slot and writes the updated state in place.

The standalone gate helper produces the negative decay gate used above:

$$
g = -\exp(A)\,\mathrm{softplus}_{\beta}(g_{\mathrm{raw}} + b),
$$

where the bias is optional and the softplus kernel switches to its linear approximation above `threshold`.

#### Chunked formulation

The prefill path computes the same recurrence in 64-token chunks. For local token indices $0 \le j \le i < C$, define the cumulative vector gate

$$
G_i = \sum_{r=0}^{i} g_r,
\qquad
E_{ij} = \exp(G_i - G_j).
$$

The implementation converts $G$ to base-2 exponent space,

$$
\widehat{G} = G\log_2(e),
\qquad
\mathrm{exp2}(\widehat{G}_i-\widehat{G}_j)=\exp(G_i-G_j),
$$

so the Triton kernels can use `exp2` without changing the mathematical decay.

The two KKT kernels construct a strictly lower-triangular correction matrix $L$ and a causal query-key matrix $P$ (`Aqk` in code):

$$
L_{ij} =
\begin{cases}
\beta_i\left\langle k_i,\;k_j\odot E_{ij}\right\rangle, & j < i,\\
0, & j \ge i,
\end{cases}
$$

$$
P_{ij} =
\begin{cases}
s\left\langle q_i,\;k_j\odot E_{ij}\right\rangle, & j \le i,\\
0, & j > i.
\end{cases}
$$

`solve_tril_kda` then computes

$$
M=(I+L)^{-1}.
$$

For a chunk matrix $K_c \in \mathbb{R}^{C\times K}$, values $V_c \in \mathbb{R}^{C\times V}$, scalar-per-token $\beta$, and the state $S_c$ at the start of the chunk, `recompute_w_u_fwd_kernel` and the chunk-state kernel compute

$$
U = M\,\mathrm{Diag}(\beta)V_c,
$$

$$
W = M\left[\mathrm{Diag}(\beta)\left(K_c\odot\exp(G)\right)\right],
$$

$$
K_c^{\star}=K_c\odot\exp(G_{C-1}-G),
\qquad
V' = U-WS_c,
$$

$$
S_{c+1}=\mathrm{Diag}(\exp(G_{C-1}))S_c+\left(K_c^{\star}\right)^\top V'.
$$

Finally, `chunk_gla_fwd_kernel_o` combines the contribution from the state before the chunk with the causal contribution produced inside the chunk:

$$
O_c=s\left(Q_c\odot\exp(G)\right)S_c+PV'.
$$

This decomposition keeps the sequential recurrence only at chunk boundaries; all token interactions within a chunk are expressed as tiled matrix operations suitable for Ascend NPU execution. `cu_seqlens`, `chunk_indices`, and `chunk_offsets` preserve the same flow for flattened variable-length batches without allowing a chunk to cross a sequence boundary.

#### Precision and layout notes

- The default chunk size is 64 tokens. Pairwise matrices are built from 16-token sub-blocks, and the triangular inverse is assembled from 16x16 blocks into the final 64x64 inverse.
- Cumulative gates, $L$, and `Aqk` are produced in fp32. The solved inverse is converted to the input dtype for the following tiled matrix multiplications.
- Chunk-boundary state is accumulated and returned in fp32. In particular, the precision-sensitive $WS_c$ residual uses fp32 operands and `input_precision="ieee"` before subtracting it from $U$.
- The physical state layout is `[V, K]`, while the equations use `[K, V]`. This avoids an extra transpose in the value-tiled state update and output kernels.

### How was this patch tested?

Added e2e nightly precision tests under `tests/e2e/nightly/single_node/ops/singlecard_ops/triton/`. Each test compares the Triton kernel against a naive float32 recurrent KDA reference ported from FLA's `naive.py`, with an RMSE-ratio tolerance of `0.005` for both output `o` and final state `ht`.

- **`test_chunk_kda_npu.py`** – Tests `chunk_kda` prefill/chunked execution across variable-length layouts using `cu_seqlens`, including:
  - Single- and multi-sequence inputs
  - Unaligned sequence lengths
  - Long sequences up to 8192 tokens
  - `H ∈ {32, 64}` and `D = 128`
  - fp16 and bf16

- **`test_fused_recurrent_kda_npu.py`** – Tests `fused_recurrent_kda` decode execution, including:
  - Variable-length recurrent mode
  - Single-token decode and short multi-token sequences
  - fp16 and bf16
  - In-place state updates with `ssm_state_indices`
  - Paged-state usage with `N ∈ {1, 4, 16}`

Run with:

```bash
pytest -sv \
  tests/e2e/nightly/single_node/ops/singlecard_ops/triton/test_chunk_kda_npu.py

pytest -sv \
  tests/e2e/nightly/single_node/ops/singlecard_ops/triton/test_fused_recurrent_kda_npu.py
```

- vLLM version: v0.23.0
- vLLM main: https://github.com/vllm-project/vllm/commit/967c5c3bc38891f4465d3f4e99917ed837bb3833

### CODE REVIEW

File: vllm_ascend/ops/triton/fla/chunk_delta_h.py [hunk 1]
```diff
@@ -242,3 +249,319 @@ def grid(meta):
...
+    {
+        "USE_G": lambda args: args["g"] is not None,
+        "USE_GK": lambda args: args["gk"] is not None,
+        "USE_INITIAL_STATE": lambda args: args["h0"] is not None,
+        "STORE_FINAL_STATE": lambda args: args["ht"] is not None,
+        "SAVE_NEW_VALUE": lambda args: args["v_new"] is not None,
+        "IS_VARLEN": lambda args: args["cu_seqlens"] is not None,
+    }
+)
+# @triton.autotune(
```
Comment (ChenxiQ) [inline vllm_ascend/ops/triton/fla/chunk_delta_h.py]: Remove comment code

File: vllm_ascend/ops/triton/fla/index_kda.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+# SPDX-FileCopyrightText: Songlin Yang, Yu Zhang
+#
+# This file contains code copied from the flash-linear-attention project.
+# The original source code was licensed under the MIT license and included
+# the following copyright notice:
+# Copyright (c) 2023-2025, Songlin Yang, Yu Zhang
```
Comment (ChenxiQ) [inline vllm_ascend/ops/triton/fla/index_kda.py]: Update copyright

File: vllm_ascend/ops/triton/fla/l2norm.py [hunk 3]
```diff
@@ -7,12 +7,18 @@
...
+
+import os
 
 import torch
 from vllm.triton_utils import tl, triton
 
 from vllm_ascend.ops.triton.triton_utils import get_vectorcore_num
 
+BT_LIST = [8, 16, 32, 64, 128]
+USE_DEFAULT_FLA_NORM = int(os.getenv("USE_DEFAULT_FLA_NORM", "0"))
```
Comment (ChenxiQ) [inline vllm_ascend/ops/triton/fla/l2norm.py]: We should not add new envs.

File: vllm_ascend/ops/triton/fla/l2norm.py [hunk 4]
```diff
@@ -40,27 +170,18 @@ def l2norm_fwd(x: torch.Tensor, eps: float = 1e-6, output_dtype: torch.dtype | N
...
-        Y=y,
-        eps=eps,
-        M=T,
-        N=D,
-        MBLOCK=MBLOCK,
-        NUM_CHUNKS=num_sub_blocks,
-    )
+    if backend == "ascend":
+        _l2norm_fwd_ascend(x, y, eps)
+    elif backend == "fla":
```
Comment (ChenxiQ) [inline vllm_ascend/ops/triton/fla/l2norm.py]: needs to double check here, why we have "fla" backend

File: vllm_ascend/ops/triton/fla/utils.py [hunk 5]
```diff
@@ -130,3 +174,79 @@ def clear_ssm_states(ssm_states: torch.Tensor, has_initial_state: torch.Tensor)
...
+    try:
+        return [
+            triton.runtime.driver.active.utils.get_device_properties(i)["max_shared_mem"]
+            for i in range(device_torch_lib.device_count())
+        ]
+    except BaseException:
+        return [-1]
+
+
+class Backend(Enum):
```
Comment (ChenxiQ) [inline vllm_ascend/ops/triton/fla/utils.py]: same here, should be removed

**Comment by ZT-AIA** [comment]
When considering the use of the `get_vectorcore_num` function in triton grid , it can help avoid performance losses caused by repeated issuance.

**Comment by ZT-AIA** [comment]
1. The difference between the triton operator without KDA in the upper part of the warehouse and the triton operator with KDA, for example, `solve_tril_16x16_kernel_kda` and `solve_tril_16x16_kernel`.

2. Please confirm whether the test cases are complete.

**Review by ZT-AIA — APPROVED** [review]

**Review by realliujiaxu — APPROVED** [review]
